File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/abstract_factory/lab/Garden.m
% Garden.m

classdef Garden < handle
	properties
		LettuceType % int
		TomatoType % int
	end

	methods
		% constructor:
		function theGarden = Garden(aLettuceType, aTomatoType)
			theGarden.LettuceType = aLettuceType;
			theGarden.TomatoType = aTomatoType;
		end

		function set.LettuceType(theGarden, aLettuceType)
			if aLettuceType ~= 1 && aLettuceType ~= 2
				error('A Garden must have a valid Lettuce type.');
			end

			theGarden.LettuceType = aLettuceType;
		end

		function set.TomatoType(theGarden, aTomatoType)
			if aTomatoType ~= 1 && aTomatoType ~= 2
				error('A Garden must have a valid Tomato type.');
			end

			theGarden.TomatoType = aTomatoType;
		end

		function result = makeLettuce(theGarden)
			if theGarden.LettuceType == 1
				result = ConventionalLettuce(1, 2);
			elseif theGarden.LettuceType == 2
				result = OrganicLettuce(1, 2);
			else
				error('Oops, should never reach this point.');
			end
		end

		function result = makeTomato(theGarden)
			if theGarden.TomatoType == 1
				result = ConventionalTomato(1, 2);
			elseif theGarden.TomatoType == 2
				result = OrganicTomato(1, 2);
			else
				error('Oops, should never reach this point.');
			end
		end
	end
end

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/abstract_factory/lab/garden.cpp
// garden.cpp

#include "conventional_lettuce.h"
#include "conventional_tomato.h"
#include "dp_exception.h"
#include "lettuce.h"
#include "organic_lettuce.h"
#include "organic_tomato.h"
#include "garden.h"
#include "tomato.h"

garden::garden(lettuce* a_lettuce_prototype)
{
	set_lettuce_prototype(a_lettuce_prototype);
}

void garden::set_lettuce_prototype(lettuce* a_lettuce_prototype)
{
	if (a_lettuce_prototype == nullptr)
	{
		throw dp_exception("void garden::set_lettuce_prototype(int a_lettuce_type) -- A garden must have a valid lettuce type.");
	}

	my_lettuce = a_lettuce_prototype->clone();
}

lettuce* garden::make_lettuce() const
{
    return my_lettuce;
}


